pkg/storage/storage_fs.go

```diff
@@ -0,0 +1,1333 @@
+package storage
+
+import (
+	"bytes"
+	"context"
+	"crypto/sha256"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/anuvu/zot/errors"
+	zlog "github.com/anuvu/zot/pkg/log"
+	apexlog "github.com/apex/log"
+	guuid "github.com/gofrs/uuid"
+	godigest "github.com/opencontainers/go-digest"
+	ispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/opencontainers/umoci"
+	"github.com/opencontainers/umoci/oci/casext"
+	"github.com/rs/zerolog"
+)
+
+const (
+	// BlobUploadDir defines the upload directory for blob uploads.
+	BlobUploadDir = ".uploads"
+	schemaVersion = 2
+	gcDelay       = 1 * time.Hour
+)
+
+// BlobUpload models and upload request.
+type BlobUpload struct {
+	StoreName string
+	ID        string
+}
+
+type StoreController struct {
+	DefaultStore ImageStore
+	SubStore     map[string]ImageStore
+}
+
+// ImageStoreFS provides the image storage operations.
+type ImageStoreFS struct {
+	rootDir     string
+	lock        *sync.RWMutex
+	blobUploads map[string]BlobUpload
+	cache       *Cache
+	gc          bool
+	dedupe      bool
+	log         zerolog.Logger
+}
+
+func (is *ImageStoreFS) RootDir() string {
+	return is.rootDir
+}
+
+func (is *ImageStoreFS) DirExists(d string) bool {
+	fi, err := os.Stat(d)
+	if err != nil && os.IsNotExist(err) {
+		return false
+	}
+
+	if !fi.IsDir() {
+		return false
+	}
+
+	return true
+}
+
+func getRoutePrefix(name string) string {
+	names := strings.SplitN(name, "/", 2)
+
+	if len(names) != 2 { // nolint: gomnd
+		// it means route is of global storage e.g "centos:latest"
+		if len(names) == 1 {
+			return "/"
+		}
+	}
+
+	return fmt.Sprintf("/%s", names[0])
+}
+
+func (sc StoreController) GetImageStore(name string) ImageStore {
+	if sc.SubStore != nil {
+		// SubStore is being provided, now we need to find equivalent image store and this will be found by splitting name
+		prefixName := getRoutePrefix(name)
+
+		imgStore, ok := sc.SubStore[prefixName]
+		if !ok {
+			imgStore = sc.DefaultStore
+		}
+
+		return imgStore
+	}
+
+	return sc.DefaultStore
+}
+
+// NewImageStoreFS returns a new image store backed by a file storage.
+func NewImageStoreFS(rootDir string, gc bool, dedupe bool, log zlog.Logger) ImageStore {
```

Comment:
Why not just call this NewImageStore()? Understood, ImageStore is now an interface

Reply:
Because there's going to be a similar method for S3, and it should be clear which is for the currently supported FS and which is for S3.

Reply:
zot.NewImageStore() and s3.NewImageStore() should be two different things, right?